x-pack/plugins/cases/server/client/metrics/alerts/aggregations/hosts.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { get } from 'lodash';
+
+import type * as estypes from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import { AggregationBuilder, AggregationResponse } from './types';
+
+export class Hosts implements AggregationBuilder {
+  constructor(private readonly uniqueValuesLimit: number = 10) {}
+
+  build() {
+    const topHits: estypes.AggregationsAggregationContainer = {
+      aggs: {
+        top_fields: {
+          top_hits: {
+            docvalue_fields: [hostName],
+            sort: [
+              {
+                '@timestamp': {
+                  order: 'desc',
+                },
+              },
+            ],
+            size: 1,
+          },
+        },
+      },
+    };
+
+    return {
+      hosts_frequency: {
+        terms: {
+          field: hostId,
+          size: this.uniqueValuesLimit,
+        },
+        ...topHits,
+      },
+      hosts_total: {
+        cardinality: {
+          field: hostId,
+        },
+      },
+    };
+  }
+
+  formatResponse(aggregations: AggregationResponse) {
+    const aggs = aggregations as HostsAggregate;
+
+    const topFrequentHosts = aggs?.hosts_frequency?.buckets.map((bucket) => ({
+      name: Hosts.getName(bucket),
+      id: bucket.key,
+      count: bucket.doc_count,
+    }));
+
+    const totalHosts = aggs?.hosts_total?.value;
+
+    const hostFields =
+      topFrequentHosts && totalHosts
+        ? { total: totalHosts, values: topFrequentHosts }
+        : { total: 0, values: [] };
+
+    return { alerts: { hosts: hostFields } };
+  }
+
+  private static getName(bucket: FieldAggregateBucket) {
+    const unsafeHostName = get(bucket.top_fields.hits.hits[0].fields, hostName);
+
+    if (Array.isArray(unsafeHostName) && unsafeHostName.length > 0) {
+      return unsafeHostName[0];
+    }
+    return unsafeHostName;
```

Comment:
Can this be `undefined`?

Reply:
Yep it can be, `lodash.get` could return `undefined`. I believe the `Array.isArray` will guard the `.length` check and then the metrics response allows `name` to be `undefined` https://github.com/elastic/kibana/pull/120982/files/f5932dd75fb55282dfe5d41f2492f250055dc279#diff-ba2d4ac2973bd7be2ba8b3ae85473e6870ce973584970ff68fda4adc64816e21R24

Reply:
I'll add it in the return type though so it's more clear 👍

Reply:
Yeah I think is good to have the return type so TS can catch it when we use the function. Thanks!